### PR TITLE
split update methods into two for offer data

### PR DESCRIPTION
**Pull request by VinozzZ**
What: 
Split the `Update` method into `Redeem` and `Finish` to fix concurrent update issue
Why:
@egonelbre mentioned in the #2031 that the current update methods may have concurrency issue and suggested that we should split the update methods. 

Thanks for submitting a PR!

## Code Review Checklist (to be filled out by reviewer)
 - [ ] Does the PR describe what changes are being made?
 - [ ] Does the PR describe why the changes are being made?
 - [ ] Does the code follow [our style guide](https://github.com/storj/docs/blob/master/code/Style.md)?
 - [ ] Does the code follow [our testing guide](https://github.com/storj/docs/blob/master/code/Testing.md)?
 - [ ] Could the PR be broken into smaller PRs?
 - [ ] Does the new code have enough tests? (*every* PR should have tests or justification otherwise. Bug-fix PRs especially)
 - [ ] Does the new code have enough documentation that answers "how do I use it?" and "what does it do?"? (both source documentation and [higher level](https://github.com/storj/docs), diagrams?)
 - [ ] Does any documentation need updating?

### CODE REVIEW

File: satellite/satellitedb/offers.go [hunk 1]
```diff
@@ -105,14 +105,26 @@ func (offers *offers) Create(ctx context.Context, o *marketing.NewOffer) (*marke
 }
 
 // Update modifies an offer entry's status and amount of offers redeemed based on offer id
```
Comment (coyle) [inline satellite/satellitedb/offers.go]: this comment needs to be updated to match the new function name

File: satellite/satellitedb/offers.go [hunk 2]
```diff
@@ -105,14 +105,26 @@ func (offers *offers) Create(ctx context.Context, o *marketing.NewOffer) (*marke
...
-func (offers *offers) Update(ctx context.Context, o *marketing.UpdateOffer) error {
+func (offers *offers) Redeem(ctx context.Context, oID int) error {
```
Comment (coyle) [inline satellite/satellitedb/offers.go]: should this be in a transaction ?

File: satellite/marketing/service.go [hunk 3]
```diff
@@ -78,14 +78,27 @@ func (s *Service) InsertNewOffer(ctx context.Context, offer *NewOffer) (o *Offer
...
+	if err != nil {
+		return Error.Wrap(err)
 	}
-	err = s.db.Offers().Update(ctx, offer)
+
+	return nil
+}
+
+// FinishOffer updates an active offer's status to be Done and its expiration time to be now
+func (s *Service) FinishOffer(ctx context.Context, oID int) (err error) {
```
Comment (coyle) [inline satellite/marketing/service.go]: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: satellite/marketing/service.go [hunk 4]
```diff
@@ -78,14 +78,27 @@ func (s *Service) InsertNewOffer(ctx context.Context, offer *NewOffer) (o *Offer
 	return o, nil
 }
 
-// UpdateOffer modifies an existing offer in the db when the offer status is set to NoStatus
-func (s *Service) UpdateOffer(ctx context.Context, offer *UpdateOffer) (err error) {
+// RedeemOffer adds 1 to the number of redeemed for an offer
+func (s *Service) RedeemOffer(ctx context.Context, uo *UpdateOffer) (err error) {
```
Comment (coyle) [inline satellite/marketing/service.go]: https://github.com/golang/go/wiki/CodeReviewComments#named-result-parameters

File: satellite/marketing/offers_test.go [hunk 5]
```diff
@@ -59,12 +59,14 @@ func TestOffer_Database(t *testing.T) {
...
-				ExpiresAt:   time.Now(),
+				ID:        new.ID,
+				Status:    marketing.Done,
+				ExpiresAt: time.Now(),
 			}
-			err = planet.Satellites[0].DB.Marketing().Offers().Update(ctx, update)
+			err = planet.Satellites[0].DB.Marketing().Offers().Redeem(ctx, update.ID)
+			require.NoError(t, err)
+
+			err = planet.Satellites[0].DB.Marketing().Offers().Finish(ctx, update.ID)
```
Comment (coyle) [inline satellite/marketing/offers_test.go]: Are you checking that the it is doing the correct thing?

File: satellite/marketing/service.go (hunk 4 above)
Comment (VinozzZ) [reply satellite/marketing/service.go]: Thanks for sharing the link. I used the named result bc I needed to pass the `error` into the monkit and that has been the patter for doing so. I can definitely change it

File: satellite/marketing/service.go (hunk 3 above)
Comment (VinozzZ) [reply satellite/marketing/service.go]: Thanks for sharing the link. I used the named result bc I needed to pass the `error` into the monkit and that has been the patter for doing so. I can definitely change it

File: satellite/marketing/offers_test.go (hunk 5 above)
Comment (VinozzZ) [reply satellite/marketing/offers_test.go]: Good point! I will add the check

File: satellite/satellitedb/offers.go [hunk 6]
```diff
@@ -118,6 +118,7 @@ func (offers *offers) Redeem(ctx context.Context, oID int) error {
 	return nil
```
Comment (fuskovic) [inline satellite/satellitedb/offers.go]: Should we create a helper function that checks if `num_redeemed` < `redeemable_cap` before executing the query statement?

Comment (VinozzZ) [reply satellite/satellitedb/offers.go]: The check is done in the query

File: satellite/marketing/offers_test.go (hunk 5 above)
Comment (coyle) [reply satellite/marketing/offers_test.go]: I think this is a good candidate for table driven tests. Not necessary but would be a good learning exercise.

https://github.com/golang/go/wiki/TableDrivenTests

File: satellite/marketing/service.go (hunk 4 above)
Comment (coyle) [reply satellite/marketing/service.go]: no need to change I forget monkit's oddities

**Pull request merged**